File: 155-min-stack/min-stack.java
class MinStack {
    Stack<Pair> st = new Stack<>();
    public MinStack() {
        
    }
    
    public void push(int val) {
        if(st.empty()){
            st.push(new Pair(val, val)); 
        }else{
            st.push(new Pair(val,min(val,st.peek().second)));
        }
        
    }
    
    public void pop() {
        if (!st.isEmpty()) {
            st.pop();
        }

        
    }
    
    public int top() {
        return st.peek().first;
        
    }
    
    public int getMin() {
        return st.peek().second;
        
    }
    public int min(int v1,int v2){
        if(v1<v2){
            return v1;
        }
        return v2;
    }
}
class Pair {
    int first, second;
    Pair(int first, int second) {
        this.first = first;
        this.second = second;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155-min-stack/min-stack.c

#define MAX 30000

typedef struct {
    int stack[MAX];
    int minStack[MAX];
    int top;
} MinStack;

/* Initialize stack */
MinStack* minStackCreate() {
    MinStack* obj = (MinStack*)malloc(sizeof(MinStack));
    obj->top = -1;
    return obj;
}

/* Push element */
void minStackPush(MinStack* obj, int val) {
    obj->top++;
    obj->stack[obj->top] = val;

    if (obj->top == 0)
        obj->minStack[obj->top] = val;
    else {
        int currentMin = obj->minStack[obj->top - 1];
        obj->minStack[obj->top] = (val < currentMin) ? val : currentMin;
    }
}

/* Pop element */
void minStackPop(MinStack* obj) {
    obj->top--;
}

/* Get top element */
int minStackTop(MinStack* obj) {
    return obj->stack[obj->top];
}

/* Get minimum element */
int minStackGetMin(MinStack* obj) {
    return obj->minStack[obj->top];
}

/* Free memory */
void minStackFree(MinStack* obj) {
    free(obj);
}



/**
 * Your MinStack struct will be instantiated and called as such:
 * MinStack* obj = minStackCreate();
 * minStackPush(obj, val);
 
 * minStackPop(obj);
 
 * int param_3 = minStackTop(obj);
 
 * int param_4 = minStackGetMin(obj);
 
 * minStackFree(obj);
*/